optimize performance, a few formatting, a bit more comments

MrZ626 opened the pull request:
Hello! I'm a indie game developer using lua and occasionally found this repo. I noticed many things to improve, so I made this PR to optimize the performance and remove some redundant codes.
All tests in **qrtest.lua** passed, but the lua version support testing seems not working, I think it has expired for a long time?
Scroll below to see benchmark result, this patch could improve performance by 10~70% based on lua version

alerque reviewed (changes requested):
I'm not the maintainer of this project, but this doesn't seem right at all. If it was on one of my projects I'd likely dismiss it entirely. There may or may not be legitimate fixes and improvements in here, but it's hard to tell. A large number of the changes are purely style/cosmetics, and many of those are for the worse. They clearly are different coding style and messes up a lot of otherwise clearly intentionally formatted bits. Some of the changes to e.g. math operators even look like they might be outright wrong. I would suggest splitting this up into individual commits with style and function changes separately with rationals for each change in the commit messages.

Also if this is in any way LLM generated I'd declare that too. It seems like a lot of senseless twiddling.

MrZ626 commented:
oh okay, I'll separate them to several commits later and remove unecessary formatting
all changes are hand-written, not AI generated

pgundlach commented:
Besides the comments from @alerque (which I totally agree with), it would like to see how much performance optimisation this code does. 

This would be a good start for benchmarking, although I don't know how to do benchmarking in Lua.

alerque commented:
I've had really mixed luck with running a profiler *inside* Lua scripts, they seem really susceptible to noisy results. I would suggest writing a Lua script that just does a massive loop generating codes in memory using loop iteration data to vary the contents a little bit each time. Then use `hyprfine` to run test scripts. Make sure you setup the tests of new and old versions to run in the same test run, not doing one, changing code around, then doing the other. That way you can be much more sure that system background stuff isn't affecting test runs differently for one set of tests or the other.

MrZ626 committed:
rewrite `bit_xor` function, saving memory, garbage cache generating, and probably improving performance

MrZ626 committed:
rewrite `binary` function, less garbage memory generated and improve performance

MrZ626 committed:
optimize `encode_string_xxx` functions, less code and less RAM usage

MrZ626 committed:
simplify `add_pad_data` function

MrZ626 committed:
optimize `add_data_to_matrix` `convert_bitstring_to_bytes` function, less RAM usage

MrZ626 committed:
several expression/logic optimizations with less code

MrZ626 committed:
simplify `add_alignment_pattern` function, remove 27 lines of hardcoding

MrZ626 committed:
replace math.fmod with % operator

MrZ626 committed:
optimize `get_pixel_with_mask` function, using number instead of string

MrZ626 committed:
cache math&string functions for better performance

MrZ626 committed:
unify spacing between paragraphs

MrZ626 committed:
add a few more comments to the main function for skipping read everything above

pgundlach commented:
For OP: 

This code has its main focus on readability, so for example changing

```
x = x - 1
y = y - 1
```

to 

```
x, y = x - 1, y - 1
```

might help performance (which I doubt, but could be), does not help readability (IMO).

I know that this is a trivial case. 

If you have updated the pull request I could suggest a few more changes. For example the function `encode_string_numeric()` could be enhanced to 

```lua
local function encode_string_numeric(str)
    local parts = {}
    for i = 1, #str, 3 do
		local a = sub(str,i,i+2)
        -- #a is 1, 2, or 3, so bits are 4, 7, or 10
		parts[#parts+1] = binary(tonumber(a), #a * 3 + 1)
	end
    return table.concat(parts)
end
```

No string concatenation, explanation of the values that are passed to `binary()`.


I am not happy reducing 

```lua
local function fill_matrix_position(matrix,bitstring,x,y)
	if bitstring == "1" then
		matrix[x][y] = 2
	else
		matrix[x][y] = -2
	end
end
```

to something like this 

```lua
local function fill_matrix_position(matrix,bitstr,x,y)
	matrix[x][y] = bitstr == "1" and 2 or -2
end
```

(I see you have pushed some more commits, so I just comment this and have a look at the commits later)

MrZ626 commented:
well I've used lua for many years, did many benchmarking experiments and optized a lot of codes, so I can basically confirm that my changes are faster than original version
https://gitspartv.github.io/LuaJIT-Benchmarks/ btw here is a reference of how to optimize lua code if anyone need

MrZ626 commented:
## `v1,v2=1,2`
Readability: 1 less line of code, not harder to read
Performance: same

## `Cond and Expr1 or Expr2`
Readability: this is basically same thing with ternary operator existed in many languages, I think it's also trivial
Performance: the speed is same, but using less RAM to save the code itself

alerque commented:
> so I can basically confirm that my changes are faster than original version

An appeal to authority here isn't going to cut it, I think this should actually be benchmarked *HERE* in this project. Also with PUC Lua and LuaJIT separately.

I see things in here that don't jive with my experience at all, such as using alias being labeled as a speed optimization. I've tried to benchmark that before and was unable to show any difference, e.g. `string.sub()` and `local sub = string.sub; sub()` work out so close I couldn't prove to myself which was more efficient. One could argue style points one way or the other, but your commit claims it is a speed optimization. I'm not buying that without *showing*  it. I'm happy to be proved wrong; it would be interesting no know for other projects.

pgundlach commented:
> well I've used lua for many years, did many benchmarking experiments and optized a lot of codes, so I can basically confirm that my changes are faster than original version https://gitspartv.github.io/LuaJIT-Benchmarks/ btw here is a reference of how to optimize lua code if anyone need

While I really appreciate patches and put a lot of work into reviewing patches, "trust me bro" is not a good foundation for benchmarking. I absolutely believe you that your changes do speed up the process of QR code generation, I'd be interested in if it is worth reducing readability of the code. (A lot of your changes do help readability) And then again, I wonder why some of the lower hanging fruits are not taken into account (like string concatenation). 

Don't get me wrong, I am happy for any pull request that enhances the code, but I'd like to see that enhancement.

MrZ626 committed:
revert to use table as buffer for `encode_string_xxx` functions (faster processing huge amount of data)

MrZ626 commented:
using alias does speed up the code, and it's listed in that benchmarking website too, though there's no huge difference but it does work
hmm maybe considering "QR code generating" is not a frequently used feature, it's not that necessary to use some extreme optimization tricks, since it shorts the code a bit too much

pgundlach commented:
I ask for some benchmarking before applying these changes.

MrZ626 commented:
> "trust me bro" is not a good foundation for benchmarking

yeah I do know that, so my habit of optimization are all based on tests (I used to code lua on almost embedded device), so the worst case is only "too small improvement and not cost-effective"

MrZ626 commented:
```lua
local encodeOld=require"qrencode_old".qrcode
local encodeNew=require"qrencode".qrcode
local function checkMatEqual(a,b)
    for y=1,#a do
        for x=1,#a[y] do
            if a[y][x]~=b[y][x] then
                return false
            end
        end
    end
    return true
end

local tests={
    "1234567890",
    "0123456789qwertyuiopasdfghjklzxcvbnm",
    "你好再见天气不错",
}
for i=1,#tests do
    for ec=1,4 do
        local _,matOld=encodeOld(tests[i],4)
        local _,matNew=encodeNew(tests[i],4)
        assert(checkMatEqual(matOld,matNew),"QR encode mismatch on test "..i.." ec "..ec)

        collectgarbage()
        local t=os.clock()
        for _=1,1000 do
            encodeOld(tests[i],ec)
        end
        local dtOld=os.clock()-t
        collectgarbage()
        t=os.clock()
        for _=1,1000 do
            encodeNew(tests[i],ec)
        end
        local dtNew=os.clock()-t
        print(string.format("Test %d-%d: Improvement %.2fx",i,ec,dtOld/dtNew))
    end
end
--[[
Lua 5.4.8
Test 1-1: Improvement 1.16x
Test 1-2: Improvement 1.15x
Test 1-3: Improvement 1.16x
Test 1-4: Improvement 1.15x
Test 2-1: Improvement 1.20x
Test 2-2: Improvement 1.20x
Test 2-3: Improvement 1.19x
Test 2-4: Improvement 1.19x
Test 3-1: Improvement 1.17x
Test 3-2: Improvement 1.17x
Test 3-3: Improvement 1.18x
Test 3-4: Improvement 1.17x

Luajit (love 11.5)
Test 1-1: Improvement 1.65x
Test 1-2: Improvement 1.57x
Test 1-3: Improvement 1.57x
Test 1-4: Improvement 1.50x
Test 2-1: Improvement 1.62x
Test 2-2: Improvement 1.71x
Test 2-3: Improvement 1.53x
Test 2-4: Improvement 1.53x
Test 3-1: Improvement 1.57x
Test 3-2: Improvement 1.58x
Test 3-3: Improvement 1.54x
Test 3-4: Improvement 1.56x
]]
```
Tested on Arch Linux x86_64

pgundlach commented:
> --[[ Results:
> Test 1 EC 1: Improvement 1.54x
> Test 1 EC 2: Improvement 1.54x
> Test 1 EC 3: Improvement 1.53x

... 

these look like promising results, thank you.

MrZ626 committed:
using alias for `debugging` flag, fix a few format issues remained

MrZ626 committed:
use early return in `get_pixel_with_mask` function, add a comment for easier understanding

pgundlach commented:
I will carefully go through the changes you have provided. I will not merge any micro-optimisations, which reduce readability. But there are some small changes that seem to bring a lot of code speedup. Tests need do be written before applying the changes (even if these are simple tests). AI can help here, I guess.

Again, the patch is greatly appreciated. 

I will report my findings here and hopefully we can sort out any issues.

MrZ626 commented:
small optimisation at hotspot might still worth it if it's not too magical to understand, for example:
1. changes in `encode_string_xxx` and `fill_matrix_position` might be not important since they only run once or very a few times (below 10) during each generation, so abandoning these changes should not slow it down much
2. changes in `get_pixel_with_mask` run a massive times for each QR code like 4e3~2e5 times, so each tiny change like just using alias makes much more sense here

MrZ626 committed:
fix main function comment

alerque commented:
> using alias does speed up the code, and it's listed in that benchmarking website too, though there's no huge difference but it does work

Me == happily proved wrong.

From recollection I think I tested this in an app with lots of function calls but relatively few inside hot loops and much other work being done, so the difference was in the noise. I'll keep that difference in mind for other hot loops with calls like that in the future.

MrZ626 commented:
glad to help~ since lua is a interpreted language without compilation optimization (luajit doesn't actually change the situation), it's possible to tell at a glance about the performance, the core idea is just less **operators/functionCall** for speed & reduce **{}** and **function** for RAM usage inflation, that's basically all

pgundlach commented:
Before we merge this, I’d like to collect a few follow-up tasks and clarify the direction a bit.

In general, I’d like to keep the focus on clarity and maintainability first. Some of the changes in this patch look like micro-optimizations that slightly reduce readability without providing a clearly demonstrated benefit. Small single-digit percentage speedups are not necessarily worth the extra complexity, unless they occur in a well-identified hotspot and we can show a real impact on typical workloads.

Here’s a checklist of things I’d like to see (or discuss) before merging:

- [x] **Profiling / benchmarking script**
  - [x] Add a small script to the repository (for example in a `tools/` or `scripts/` directory) that can be used to profile / benchmark the QR code generation.
  - [x] Document briefly how to run it (e.g. in a comment at the top of the script or a short README section).

- [x] **Tests for non-trivial functions**
  - [ ] Add tests for the non-trivial functions that have been changed.
  - [ ] For each such function, verify that the new implementation produces the same results as the original implementation (for a representative set of inputs).
  - [ ] Make sure these tests are easy to run as part of the existing test setup.

- [x] **Hotspot analysis**
  - [ ] Use the profiling script to identify real hotspots in typical usage scenarios.
  - [ ] Summarize which functions actually dominate the runtime and where optimizations would have the biggest impact.
  - [ ] (Optional but nice) Attach or reference profiling output so we can revisit decisions later.

- [ ] **Revisit micro-optimizations**
  - [ ] Go through the purely “micro” changes that primarily affect readability (e.g., less clear expressions, more complex code paths).
  - [ ] Remove or simplify the ones that do not show a clear benefit in the profiling results.
  - [ ] Keep optimizations that demonstrably improve performance in hotspots without making the code significantly harder to understand.

- [ ] **Documentation / comments**
  - [ ] Where we keep non-obvious optimizations, add short comments explaining why they are there and what kind of performance gain they provide (or in which scenario they matter).
  - [ ] Ensure that future contributors can understand the intent without having to reverse-engineer the benchmark setup.

I really appreciate the work that went into this patch. With some profiling data and a bit of cleanup around readability vs. micro-optimizations, I think we can end up with a version that’s both fast and easy to maintain.

One practical note: I’d prefer if we avoid force-pushing to this branch while we’re working through the review, so that it stays easy to follow the history and comments. I might also contribute a few changes directly to this pull request, but I can’t promise how quickly I’ll be able to do that.

MrZ626 commented:
I'm worrying we are making things a bit too formal and too complicated... is there any actual comment or suggestion about any block of my changes?
I didn't add any new algorithm, and I don't think new complexity is introduced
if you guys really worry I might break something, I could drop this PR and make a new one with only aliasing and a few most important simplifying on hotspot, that will take less time for you to review and understand

pgundlach commented:
> I'm worrying we are making things a bit too formal and too complicated...

I feel your sentiments, and I promise I will try / help to make things as easy as possible. I use qrcode.lua in production and I don't want any regressions. This seems to be a good opportunity to add some more tests. Your code looks fine, these are purely for adding some more optimisation.

I will add an ad hoc implementation of a profiler and add some tests if this is okay with you. I know that the profiler is not good, but adds a little insight which functions are called most often.

pgundlach committed:
Add profiling script

pgundlach committed:
Add some tests

pgundlach committed:
Fix luacheck warnings, remove test cache

pgundlach committed:
Add modern gh-actions-lua

MrZ626 committed:
use HEX instead of OCT in `binary` function, for better performance

pgundlach committed:
Enhance Readme, remove debugging switch, spelling

MrZ626 committed:
simplify formula of mask 110 & 111

pgundlach committed:
Replace string concatenation by table.concat

This does not improve anything, it is just not clean code

pgundlach committed:
Optimize bit_xor for portable fast path

Use native (~) when available and fall back to a prebuilt 256x256 XOR
lookup table with a portable initializer. Keeps the hot path O(1) on all
Lua versions without relying on bit32/bit. Updated comments accordingly.

pgundlach commented:
I have an implementation here which replaces `bit_xor()` and significantly speeds up the qr code generation. I hope it is okay with you when I push this to the branch. If not, we can revert this to the current implementation. So far the tests pass, but in the end we should make sure the QR codes don't change.

MrZ626 commented:
oh you use the ~ operator, but in earlier version of lua this will crash the whole file, it's syntax error, not runtime error
you can try to use bit.bxor (luajit feature) in this way tho, it's not new syntax

pgundlach committed:
Remove native ~ check and rely on lookup XOR

Always use the portable arithmetic XOR to populate the 256x256 lookup
table, avoiding errors on Lua versions without native bitwise operators.

pgundlach commented:
> oh you use the ~ operator, but in earlier version of lua this will crash the whole file, it's syntax error, not runtime error

I see, the tests failed on the earlier version. I have reverted the commit and always use the slow_xor function to pre-build the xor table.

MrZ626 commented:
I'm not sure if `load` & `loadstring` are allowed, if yes, you can test if ~ operator is available with something like this:
```lua
xor=loadstring("return function(a,b) return a ~ b end")
```
(picking `load` or `loadstring` depends on lua version, I forgot the detail, anyway this idea works)

pgundlach commented:
If we stop here (without loadstring something like that), we already have a huge win:

### Benchmark: 1000 QR codes, value `"The quick brown fox jumps over the lazy dog."`

| Version                           | Time (s) | Speedup vs. old version |
|-----------------------------------|---------:|------------------------:|
| Old version                       | 7.07     | –                       |
| New version (without XOR table)  | 6.14     | ≈ 13.2% faster          |
| New version (with XOR table)     | 4.58     | ≈ 35.2% faster          |

Additional comparison:

- Speedup of XOR-table version vs. new version without table: **≈ 25.4% faster**

These benchmarks are just some ad-hoc benchmarks on my laptop, no scientific research.

I think we are almost there. Next step would be to write a small black box test (for example a file with a hundred strings and a QR-matrix representation of these strings) and then test these against both implementations. I don't expect any changes. 
I am not sure that I can make this tomorrow.

pgundlach committed:
Add black box tests

MrZ626 commented:
Another way is just keeping a copy of old version, and compared the outputs with two versions so you can generate random test case easily
I was testing my version in this way, got exactly same output, and reached 2.2x speed just now (won't merge since there are many aggressive optimizations)

pgundlach commented:
I am happy with the current status of the pull request, I think it could be merged now. There are certainly more optimisations possible (for example with the xor table), but IMO everything is fine. Your opinion @MrZ626 ?

MrZ626 commented:
let me check if there's any safe changes could be merged here, wait for a moment

pgundlach commented:
No hurry, It is late here, I go to bed now ;-)

MrZ626 committed:
remove function call in mask 100 formula

MrZ626 committed:
optimize data concatenating in `arrange_codewords_and_calculate_ec` function
WARN: logic might be changed, but it passed all of my local tests

MrZ626 committed:
Optimize unecessary table creations in `calculate_error_correction` function.
Reason: two `convert_to_xxx` functions created tons of garbage RAM and this can be avoided, improve around 30~40% performance (in LuaJIT)
Note: All my tests passed, but could be abandon if you really concern it a lot.
WARN: though logic is unchanged, codes are simplified a lot, need review carefully.

MrZ626 commented:
The last commit **can be abandoned** if you really concerned about it. more details in the commit message

MrZ626 committed:
Revert "optimize data concatenating in `arrange_codewords_and_calculate_ec` function"

This reverts commit 3cd9ca11337da6747538f3670c9627debfcfdb8e.

MrZ626 committed:
fix variable name

MrZ626 committed:
remove the redundant math.ceil

pgundlach commented:
> The last commit **can be abandoned** if you really concerned about it. more details in the commit message

All tests pass, so I think it is okay. I will take a look but most likely we could merge once you give your okay.

pgundlach committed:
Translate comments into Englisch

MrZ626 commented:
about the black box test, I think we need some large-size test cases, some issues like multi-block feature may only happen on high versions (I failed on some cases earlier)

pgundlach committed:
Separate data and logic in qrblackbox, generate

Run

lua qrblackbox.lua --generate "myteststring"

to create another test case for qrblackbox

pgundlach commented:
> about the black box test, I think we need some large-size test cases, some issues like multi-block feature may only happen on high versions (I failed on some cases earlier)

I have split qrblackbox into two pieces, test data and logic. With `lua qrblackbox.lua --generate "test data"` you can create tests for qrblackbox.

If you have good test cases, please add them.

pgundlach committed:
Describe qrblackbox --generate in README.md

MrZ626 committed:
optimize `get_pixel_with_mask` function, rewrite `add_data_to_matrix` function with a simpler zig-zag filling implemention
remove separated `get_next_free_positions` function

MrZ626 committed:
add a middle-size test case (generated by old version, can be correctly scanned)

MrZ626 commented:
The old zig-zag filling algorithm is the last thing which worth optimizing I think
Reason: doesn't create any table objects, more straight-forward to understand (with comments, should be easy enough to understand)
Note 1: it's not based on 8-bit batch. I don't know why old implementation did this byte by byte, the whole bitstream is already fixed at this time so there's no reason to slice it
Note 2: I removed the code for testing it. If it breaks, it will fail on the next `add_data_to_matrix` test, so it's not necessary

pgundlach commented:
looks good to me.

MrZ626 commented:
and I'm done now, could be merged if you think it's safe enough for production environment
(I still worry a tiny bit since we don't have dynamic test cases on larger scale, my local test script genarates random huge case for oldest & newest version and assert the results are same, which feels more reliable)

MrZ626 committed:
whitespace fix, satisfying the luachecker

pgundlach committed:
Add some more tests

pgundlach commented:
I have added a few more tests in the blackbox file. 

I think this is a really good pull request, great speedup, good testing and profiling code (although profiling is just basic) and good documentation. Before I merge, I will wait a few hours (perhaps some more things come into our minds) but this is IMO a really good patch. Thank you for bringing this up and being patient with me.

Here is my suggestion for a commit message, feel free to update or correct it.

Improve QR encoding performance
- inline mask pixel lookup to reduce per-bit overhead
- rewrite add_data_to_matrix with a simpler zig-zag fill and drop get_next_free_positions
- keeps behavior identical while speeding up matrix generation
- Basic unit and black-box tests for core QR routines; run with `make test`.
- Lightweight profiler for spotting hotspots in typical encodes via `make profile`.
- Bitwise XOR uses a precomputed 256x256 lookup table to stay fast and portable across Lua versions.
- README refreshed with contributor-friendly development notes.

pgundlach commented:
Done. Thanks again, extremely helpful!
